In [5]:
#Check for file paths 

In [6]:
!ls ./../

README.md        Time_Complexity  csvfunction.py   stockprices.txt
Space_Complexity __pycache__      stock_prices.csv txtfunction.py


In [ ]:
import csv 

stock_prices = []

with open('./../stock_prices.csv', 'r') as file:
    reader = csv.reader(file)
    next(reader) #skip the header row 
    for row in reader:
        day = row[0]
        price = float(row[1])
        stock_prices.append([day, price])

#print(stock_prices)

#### ----------------------------Alternatively-------------------------

In [22]:
stock_prices = []

with open ('./../stock_prices.csv', 'r') as file:
    next(file) #skip the header row
    for line in file:
        tokens = line.strip().split(',')
        day = tokens[0]
        price = float(tokens[1])
        stock_prices.append([day, price])

#print(stock_prices)

In [24]:
#this is a 2Dimensional array where the the first element in each list is [0] - that day and the price [1]
stock_prices

[['6-Mar', 310.0],
 ['7-Mar', 340.0],
 ['8-Mar', 380.0],
 ['9-Mar', 302.0],
 ['10-Mar', 297.0],
 ['11-Mar', 323.0]]

In [25]:
#if we want to check the price of stock price for march 9

for element in stock_prices:
    if element[0] == '9-Mar':
        print(element[1])

302.0


## Big O Analysis:
- **O(n) LIST:** - So in the example above, we iterated through the stock prices using the **list** before we found the **March 9** stock price. Imagine if we had many stock prices, we will iterate through  the entire stock prices. So using a **list** would work here but it is not very efficient because the Big O complexity here is **O(n)**

- **O(1) DICTIONARY** - In python there is dictionary whicn will perform the same operation with a constant time `O(1)` operation as shown below:

- Since it is a dictionary, we can add a key which in this case is the day and set each day key equal to price 

In [26]:
stock_prices = {}

with open('./../stock_prices.csv', 'r') as file:
    next(file) # skip the header row 
    for line in file:
        tokens = line.strip().split(',')
        day = tokens[0]
        price = float(tokens[1])
        stock_prices[day] = price

stock_prices

{'6-Mar': 310.0,
 '7-Mar': 340.0,
 '8-Mar': 380.0,
 '9-Mar': 302.0,
 '10-Mar': 297.0,
 '11-Mar': 323.0}

In [28]:
#So if want to fnd the stock price for March 9 - it is very simple 
stock_prices['9-Mar']

302.0


### Hashmap or Hash Table 
* Dictionary implementation uses Hash Table or Hashmap - which is an O(1) complexity. 

- Dictionary is a python specific implementation of Hashmap or Hash Table - they are the same. So dictionary implements this operation using what is called **Hash Function**

### Hash Function 
- There are different way of implementing `Hash Function` - which is using `ASCII` numbers. For example - Assuming that we want to store and retrieve `march 6` in the dictionary, each value will be assigbed an `ASCII number` in the `ASCII Table` - for example, `m` in the `ASCII` table is encoded `109` in the `ACII Table`.


| Value | ASCII numbers |
| ---   | ---           |
| m     |  109          |
| a     |  97           |
| r     |  114          |
| c     |  99           |
| h     |  104          |
|       |  32           |
| 6     |  54           |
| SUM   |  609          |

* Like I said earlier there are different types of Hash functio, in the above example - I have only shown the `ASCII` method

**Big O Analysis:** 
- Look up by key is **O(1)** on average 
- insertion/Deletion is **O(1)** on average 

### Refrences
* [Lookup Tables](https://www.lookuptables.com/)

#### List of classes for implementing Hash Table in different Programming Language 

|       |  Class      | Code Sample                               |
| ---   | ---         | ---                                       |
| Python| dictionary  | prices = {'march 6': 310, 'march 7': 430} |
| JAVA  | HashMap     | HashMap<String, Integer> prices = new HashMap<String, Integer>(); prices.put('march 6', 310); prices.put('march 7', 430);  |
| JAVA  | LinkedHasMap| LinkedHashMap<String, Integer> prices = new LinkedHashMap<String, Integer>(); prices.put('march 6', 310); prices.put('march 7', 430);   |
| C++   | std:map     | std:map<string,int> prices; prices['march 6'] = 310; prices['march 7'] = 430 | 


 #### Implementing HashMap in Python
* First step in implementing `Hash Table` is implementing the `Hash Function`

In [37]:
def get_hash(key):
    h = 0
    for char in key:
        h += ord(char)
    return h % 100  # we are dividing by 100 assuming the 100 is the total sum 

In [38]:
# we have implemented the same Hash Table using the ASCII, for example we can call ord(m) and it will return same 109

ord('m')

109

In [39]:
#Now to use the get_hash(key)

get_hash('march 6')

9

In [40]:
get_hash('march 28')

61

In [48]:
class HashTable:
    def __init__(self) -> None:
        self.MAX = 100 
        self.arr = [None for i in range(self.MAX)]

    def get_hash(self, key):
        h = 0
        for char in key:
            h += ord(char)
        return h % self.MAX
    
    def add(self, key, val):
        h = get_hash(key)
        self.arr[h] = val

    def get(self, key):
        h = get_hash(key)
        return self.arr[h]



In [49]:
t = HashTable()
t.get_hash('march 6')
t.add('march 6', 130)
t.get('march 6')

130

In [47]:
t.arr

[None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 130,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None]

- Let's say we want to perform some sort of operation like:

`t.add('march 6', 130)`

- And retrieve this value using:

`t.get('march 6')`

### Modifying Hash Function to include `set` and `set` method 

- Now that we have implemented the `add` and `get` function into the HashMap class. Wouldn't it be nice to perform an operation like:

`t['march 6'] = 130`

- We can also implement delete method:

` del t['march 6]`


### Reference 
- Luckily for us, python have this standard operation:

- [Python Standard Operators as functions](https://docs.python.org/3/library/operator.html)

In [62]:
class HashTable:
    def __init__(self) -> None:
        self.MAX = 100 
        self.arr = [None for i in range(self.MAX)]

    def get_hash(self, key):
        h = 0
        for char in key:
            h += ord(char)
        return h % self.MAX
    
    def __setitem__(self, key, val):
        h = get_hash(key)
        self.arr[h] = val

    def __getitem__(self, key):
        h = get_hash(key)
        return self.arr[h]

    def __delitem__(self, key):
        h = get_hash(key)
        self.arr[h] = None 

In [63]:
t = HashTable()
t['march 6'] = 130
t['march 1'] = 20
t['dec 17'] = 27

In [64]:
t.arr

[None,
 None,
 None,
 None,
 20,
 None,
 None,
 None,
 None,
 130,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 27,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None]

In [55]:
t['march 6']

130

In [57]:
t['dec 17']

27

In [65]:
del t['march 6']

### Collision Handling in Hash Table | Approaches for solving collision handling
- Implement chaining in python 
- Separate chaining 
- Linear probing 

In [124]:
class HashTable:
    def __init__(self):
        self.MAX = 10
        self.arr = [[] for i in range(self.MAX)]

    def get_hash(self, key):
        hash = 0
        for char in key:
            hash += ord(char)
        return hash % self.MAX
    
    def __getitem__(self, key):
        """
        We will modify this function to iterate through the linked list, find the index and return the value
        : self.arr[h] - will give me all the hash key, so we want to iterate through all the key and check if that matches the key 
        """
        h = self.get_hash(key)
        for element in self.arr[h]:
            if element[0] == key:
                return element[1]
        
    
    def __setitem__(self, key, val):
        """
        This method will handle hashmap collision. 
        : Update -> assuming we have a situation when the value of march 6 changes and we want to update 
        : t['march 6'] = 120
        : t['march 6'] = 78

        So we will first check if the key exists and if not append the key,value
        :self.arr[h].append((key, value))
        """
        h = self.get_hash(key)
        found = False 
        for idx, element in enumerate(self.arr[h]):
            if len(element)==2 and element[0]==key:
                #element[1] = val #now this would have worked if we have used a different array instead of tuple but because tuple is immutatble - this won't work 
                self.arr[h][idx] = (key, val) # but instead we will just insert another tuple in that particular index - idx
                found = True 
                break 
        if not found:
            self.arr[h].append((key, val))   

    def __delitem__(self, key):
        h = self.get_hash(key)
        for index, element in enumerate(self.arr[h]):
            if element[0] == key:
                del self.arr[h][index]    

In [125]:
t = HashTable()

In [126]:
t["march 6"] = 120 
t["march 6"] = 78
t["march 8"] = 67
t["march 9"] = 4
t["march 17"] = 459 

In [127]:
t.arr

[[],
 [('march 8', 67)],
 [('march 9', 4)],
 [],
 [],
 [],
 [],
 [],
 [],
 [('march 6', 78), ('march 17', 459)]]

In [128]:
t["march 6"]

78

In [129]:
t['march 17']

459

In [130]:
del t['march 6']

In [131]:
t.arr

[[],
 [('march 8', 67)],
 [('march 9', 4)],
 [],
 [],
 [],
 [],
 [],
 [],
 [('march 17', 459)]]

* we are able to see that `march 6` and `march 17` have experience the same index (this is an example of hash collission)

- So suppose we store more data like below:

- So clearly we can see that `march 6` and `march 17` are demonstrating `Hash collission` - we will try to fix this by modifying the arr to `initialize an emply list for our key value pair` and also modify  the `__setitem__` function to iterate through our linked list 